File: benchmark/fusion/tupling/page1.f
Config SampleSize = 20;

Inductive Tree = leaf Int | node {Tree, Tree};
Inductive List = nil Unit | cons {Int, List};

cat = fix (
  \f: List -> List -> List. \a: List. \b: List.
  match a with
    nil _ -> b
  | cons {h, t} -> cons {h, (f t b)}
  end
);

max = \a: Int. \b: Int. if < a b then b else a;

depth = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    leaf w -> 0
  | node {l, r} -> + 1 (max (f l) (f r))
  end
);

deepest = fix (
  \f: Tree -> {List, Compress Tree}. \t: Tree.
  match t with
    leaf w -> {cons {w, nil unit}, t}
  | node {l, r} ->
    let lres = f l in
      let rres = f r in
        if > (depth lres.2) (depth rres.2) then
          {lres.1, t}
        else if == (depth lres.2) (depth rres.2) then
          {cat lres.1 rres.1, t}
        else {rres.1, t}
  end
);

main = \t: Tree. (deepest t).1;

File: ml-benchmark/fusion/tupling/page1.ml
let sampleSize = 20

type tree =
  | Leaf of int
  | Node of tree * tree

type llist =
  | Nil
  | Cons of int * llist

let rec cat a b =
  match a with
  | Nil -> b
  | Cons (h, t) -> Cons (h, cat t b)

let max a b = if a < b then b else a

let rec depth t =
  match t with
  | Leaf w -> 0
  | Node (l, r) -> 1 + max (depth l) (depth r)

let rec deepest t =
  match t with
  | Leaf w -> (Cons (w, Nil), t)
  | Node (l, r) ->
      let lres = deepest l in
      let rres = deepest r in
      let ltree =
        match lres with
        | (x, y) -> y
      in
      let rtree =
        match rres with
        | (x, y) -> y
      in
      if depth ltree > depth rtree then
        match lres with
        | (x, y) -> (x, t)
      else if depth ltree = depth rtree then
        let llist =
          match lres with
          | (x, y) -> x
        in
        let rlist =
          match rres with
          | (x, y) -> x
        in
        (cat llist rlist, t)
      else
        match rres with
        | (x, y) -> (x, t)

let program t =
  match deepest t with
  | (x, y) -> x